tools/filetop_getattr.py

```diff
@@ -0,0 +1,412 @@
+"""
+
+import os
+import sys
+
+import bcc
+
+from bcc import topclass
+
+# define BPF program
+BPF_TEXT = """
+#include <uapi/linux/ptrace.h>
+#include <linux/blkdev.h>
+#include <linux/mount.h>
+#include <linux/fs_pin.h>
+
+/*
+ * XXX: struct mount is defined in fs/mount.h, which is not included in kernel
+ * headers. So we duplicate the definition here.
+ * The implementation may change across kernel versions, which should not affect
+ * this BPF program as long as the fields that we extract information from do
+ * not change. The below is from:
+ * https://github.com/torvalds/linux/blob/v5.11/fs/mount.h#L39
+ */
+struct mount {
+    struct hlist_node mnt_hash;
+    struct mount *mnt_parent;
+    struct dentry *mnt_mountpoint;
+    struct vfsmount mnt;
+    union {
+        struct rcu_head mnt_rcu;
+        struct llist_node mnt_llist;
+    };
+#ifdef CONFIG_SMP
+    struct mnt_pcp __percpu *mnt_pcp;
+#else
+    int mnt_count;
+    int mnt_writers;
+#endif
+    struct list_head mnt_mounts;    /* list of children, anchored here */
+    struct list_head mnt_child; /* and going through their mnt_child */
+    struct list_head mnt_instance;  /* mount instance on sb->s_mounts */
+    const char *mnt_devname;    /* Name of device e.g. /dev/dsk/hda1 */
+    struct list_head mnt_list;
+    struct list_head mnt_expire;    /* link in fs-specific expiry list */
+    struct list_head mnt_share; /* circular list of shared mounts */
+    struct list_head mnt_slave_list;/* list of slave mounts */
+    struct list_head mnt_slave; /* slave list entry */
+    struct mount *mnt_master;   /* slave is on master->mnt_slave_list */
+    struct mnt_namespace *mnt_ns;   /* containing namespace */
+    struct mountpoint *mnt_mp;  /* where is it mounted */
+    union {
+        struct hlist_node mnt_mp_list;  /* list mounts with the same mountpoint */
+        struct hlist_node mnt_umount;
+    };
+    struct list_head mnt_umounting; /* list entry for umount propagation */
+#ifdef CONFIG_FSNOTIFY
+    struct fsnotify_mark_connector __rcu *mnt_fsnotify_marks;
+    __u32 mnt_fsnotify_mask;
+#endif
+    int mnt_id;         /* mount identifier */
+    int mnt_group_id;       /* peer group identifier */
+    int mnt_expiry_mark;        /* true if marked for expiry */
+    struct hlist_head mnt_pins;
+    struct hlist_head mnt_stuck_children;
+};
+
+/*
+ * Helper macro to manipulate data structures
+ */
+#ifndef offsetof
+#define offsetof(TYPE, MEMBER)  ((size_t)&((TYPE *)0)->MEMBER)
+#endif
+
+/*
+ * container_of - cast a member of a structure out to the containing structure
+ * @ptr:        the pointer to the member.
+ * @type:       the type of the container struct this is embedded in.
+ * @member:     the name of the member within the struct.
+ */
+#ifndef containerof
+#define containerof(ptr, type, member) ({                      \
+    const typeof( ((type *)0)->member ) *__mptr = (ptr);    \
+    (type *)( (char *)__mptr - offsetof(type,member) );})
+#endif
```

Comment:
I just added the above two *generic* macros in the helpers.h
https://github.com/iovisor/bcc/commit/cda7acdbddb19a5a258f881711845e41463dabbf
so you don't need to define here.

Reply:
Thank you! Will test and remove them soon.